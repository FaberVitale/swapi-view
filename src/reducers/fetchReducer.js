//@flow
import { FETCH_FAIL, FETCH_SUCCESS, FETCH_RESOURCE } from "../action/types";
import { notEqual } from "../util/functions";

export type State = {|
  +pending: Array<string>,
  +failed: { [route: string]: string }
|};

const initState: State = {
  pending: [],
  failed: {}
};

type FetchReducer = (state: State, action: Action) => State;

/* Holds info on currently pending and failed requests */
const fetchReducer: FetchReducer = (state = initState, action) => {
  if (!action.request) {
    return state;
  }
  const { type, request } = action;
  const name: string = request.name;

  switch (type) {
    case FETCH_FAIL:
      let cause = action.cause || "error";

      return {
        pending: state.pending.filter(notEqual(name)),
        failed: Object.assign({}, state.failed, {
          [name]: cause
        })
      };
    case FETCH_SUCCESS:
      return {
        pending: state.pending.filter(notEqual(name)),
        failed: state.failed
      };
    case FETCH_RESOURCE:
      let failed;
      if (Object.prototype.hasOwnProperty.call(state.failed, name)) {
        failed = { ...state.failed };

        delete failed[name];
      } else {
        failed = state.failed;
      }

      return {
        pending: [...state.pending, name],
        failed
      };
    default:
      return state;
  }
};

export default fetchReducer;
