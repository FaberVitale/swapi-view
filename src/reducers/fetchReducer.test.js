import { default as reducer } from "./fetchReducer";
import {
  fetchResourceAction,
  fetchFailAction,
  fetchSuccessAction
} from "../__mocks__/action";

describe("reducers/fetchReducer", () => {
  it("returns the initial state if undefined is passed", () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it("returns the previous state if an unused action type is used", () => {
    expect(
      reducer(reducer(undefined, {}), { type: "@@/TEST" })
    ).toMatchSnapshot();
  });

  it("adds request.name to pending if a fetchResource action is passed", () => {
    expect(
      reducer(reducer(undefined, {}), fetchResourceAction.page)
    ).toMatchSnapshot();
  });

  it("removes request.name from pending, if present, on a fetchSuccess action", () => {
    expect(
      reducer(
        reducer(reducer(undefined, {}), fetchResourceAction.page),
        fetchSuccessAction.page
      )
    ).toMatchSnapshot();
  });

  it("adds failed[action.request.name] = action.cause on fetchFailAction", () => {
    expect(reducer(reducer(undefined, {}), fetchFailAction)).toMatchSnapshot();
  });

  it("removes request.name from failed if present, on fetchResource", () => {
    const before = {
      pending: [],
      failed: {}
    };

    const after = reducer(before, fetchFailAction);
    expect(
      Object.keys(reducer(after, fetchResourceAction.page).failed)
    ).toHaveLength(0);
  });

  it('uses a generic "error" string if action.cause is falsy', () => {
    const action = Object.assign({}, fetchFailAction);

    delete action.cause;

    expect(reducer(reducer(undefined, {}), action)).toMatchSnapshot();
  });
});
