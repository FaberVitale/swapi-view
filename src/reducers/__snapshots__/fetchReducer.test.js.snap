// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/fetchReducer adds failed[action.request.name] = action.cause on fetchFailAction 1`] = `
Object {
  "failed": Object {
    "/people/page/5": "500",
  },
  "pending": Array [],
}
`;

exports[`reducers/fetchReducer adds request.name to pending if a fetchResource action is passed 1`] = `
Object {
  "failed": Object {},
  "pending": Array [
    "/people/page/5",
  ],
}
`;

exports[`reducers/fetchReducer removes request.name from pending, if present, on a fetchSuccess action 1`] = `
Object {
  "failed": Object {},
  "pending": Array [],
}
`;

exports[`reducers/fetchReducer returns the initial state if undefined is passed 1`] = `
Object {
  "failed": Object {},
  "pending": Array [],
}
`;

exports[`reducers/fetchReducer returns the previous state if an unused action type is used 1`] = `
Object {
  "failed": Object {},
  "pending": Array [],
}
`;

exports[`reducers/fetchReducer uses a generic "error" string if action.cause is falsy 1`] = `
Object {
  "failed": Object {
    "/people/page/5": "error",
  },
  "pending": Array [],
}
`;
